src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1640,52 +1669,108 @@ ExtractInsertPartitionValue(Query *query, Var *partitionColumn)
 /* RouterSelectJob builds a Job to represent a single shard select query */
 static Job *
 RouterSelectJob(Query *originalQuery, RelationRestrictionContext *restrictionContext,
-				bool *returnQueryRoutable)
+				DeferredErrorMessage **planningError)
 {
 	Job *job = NULL;
 	Task *task = NULL;
-	bool queryRoutable = false;
 	StringInfo queryString = makeStringInfo();
 	uint64 shardId = INVALID_SHARD_ID;
 	List *placementList = NIL;
 	List *relationShardList = NIL;
 	bool replacePrunedQueryWithDummy = false;
+	bool requiresMasterEvaluation = false;
+	List *rangeTableList = NIL;
+	ListCell *rangeTableCell = NULL;
+	bool hasNullifiedModification = false;
 
 	/* router planner should create task even if it deosn't hit a shard at all */
 	replacePrunedQueryWithDummy = true;
 
-	queryRoutable = RouterSelectQuery(originalQuery, restrictionContext,
-									  &placementList, &shardId, &relationShardList,
-									  replacePrunedQueryWithDummy);
+	/* check if this query requires master evaluation */
+	requiresMasterEvaluation = RequiresMasterEvaluation(originalQuery);
 
-
-	if (!queryRoutable)
+	RouterSelectQuery(originalQuery, restrictionContext,
+					  &placementList, &shardId, &relationShardList,
+					  replacePrunedQueryWithDummy, planningError);
+	if (*planningError)
 	{
-		*returnQueryRoutable = false;
 		return NULL;
 	}
 
 	job = CreateJob(originalQuery);
 
+	/* extract range table entries */
+	ExtractRangeTableEntryWalker((Node *) originalQuery, &rangeTableList);
+
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+
+		if (rangeTableEntry->rtekind == RTE_SUBQUERY &&
+			(!bms_is_empty(rangeTableEntry->updatedCols) ||
+			 (ACL_DELETE & rangeTableEntry->requiredPerms)))
+		{
+			hasNullifiedModification = true;
+			break;
+		}
+	}
+
+	/* If there are no shards for update, create an empty task list for modifications */
+	if (hasNullifiedModification)
+	{
+		job->taskList = NIL;
+		return job;
+	}
+
 	pg_get_query_def(originalQuery, queryString);
 
-	task = CreateTask(ROUTER_TASK);
+	if (originalQuery->commandType == CMD_UPDATE || originalQuery->commandType ==
+		CMD_DELETE)
+	{
+		ListCell *relationShardCell = NULL;
+		char replicationModel = REPLICATION_MODEL_INVALID;
+
+		foreach(relationShardCell, relationShardList)
+		{
+			RelationShard *relationShard = (RelationShard *) lfirst(relationShardCell);
+			DistTableCacheEntry *cacheEntry =
+				DistributedTableCacheEntry(relationShard->relationId);
+
+			if (replicationModel == REPLICATION_MODEL_INVALID)
+			{
+				replicationModel = cacheEntry->replicationModel;
+			}
+			else if (replicationModel != cacheEntry->replicationModel)
+			{
+				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								errmsg("cannot perform an UPDATE on tables with "
```

Comment:
Do we care that much? I think we only care about the replication model of the placement that is being modified.